core/storage/story/gae_models.py

```diff
@@ -0,0 +1,199 @@
+    """Storage model for the content of a story snapshot."""
+    pass
+
+
+class StoryModel(base_models.VersionedModel):
+    """Model for storing stories.
+
+    This class should only be imported by the story domain file, the
+    story services file, and the story model test file.
+    """
+    SNAPSHOT_METADATA_CLASS = StorySnapshotMetadataModel
+    SNAPSHOT_CONTENT_CLASS = StorySnapshotContentModel
+    ALLOW_REVERT = True
+
+    # The title of the story.
+    title = ndb.StringProperty(required=True, indexed=True)
+    # A high-level description of the story.
+    description = ndb.StringProperty(indexed=False)
+    # The topic id the story corresponds to.
+    topic = ndb.StringProperty(indexed=False)
+    # A set of notes, that describe the characters, main storyline, and setting.
+    notes = ndb.StringProperty(indexed=False)
+    # The schema version for the story.
+    schema_version = (
+        ndb.IntegerProperty(required=True, default=1, indexed=True))
+    # The ISO 639-1 code for the language this question is written in.
+    language_code = ndb.StringProperty(required=True, indexed=True)
+    # The story graph specifying the connection between nodes.
+    story_contents = ndb.JsonProperty(default={}, indexed=False)
+
+    def _trusted_commit(
+            self, committer_id, commit_type, commit_message, commit_cmds):
+        """Record the event to the commit log after the model commit.
+
+        Note that this extends the superclass method.
+
+        Args:
+            committer_id: str. The user_id of the user who committed the
+                change.
+            commit_type: str. The type of commit. Possible values are in
+                core.storage.base_models.COMMIT_TYPE_CHOICES.
+            commit_message: str. The commit description message.
+            commit_cmds: list(dict). A list of commands, describing changes
+                made in this model, which should give sufficient information to
+                reconstruct the commit. Each dict always contains:
+                    cmd: str. Unique command.
+                and then additional arguments for that command.
+        """
+        super(StoryModel, self)._trusted_commit(
+            committer_id, commit_type, commit_message, commit_cmds)
+
+        committer_user_settings_model = (
+            user_models.UserSettingsModel.get_by_id(committer_id))
+        committer_username = (
+            committer_user_settings_model.username
+            if committer_user_settings_model else '')
+
+        StoryCommitLogEntryModel(
+            id=('story-%s-%s' % (self.id, self.version)),
+            user_id=committer_id,
+            username=committer_username,
+            story_id=self.id,
+            commit_type=commit_type,
+            commit_message=commit_message,
+            commit_cmds=commit_cmds,
+            version=self.version
+        ).put_async()
+
+
+class StoryCommitLogEntryModel(base_models.BaseModel):
```

Comment:
So much duplication -- worth making a BaseCommitLogEntryModel subclass of base_models.BaseModel and having all these commit log entry classes subclass it?

Reply:
Yep that approach sounds right (I'll review the code in a bit).